### PR TITLE
Proposal for database schemas

### PR DESCRIPTION
**Pull request by johnolos**
We can use this PR for further discussion from a technical standpoint.
Feel free to submit commits to this PR.

work in progress

### CODE REVIEW

**Comment by johnolos** [comment]
Some notes:
I had to use a milestone of slick so that we can continue to use scala 2.12. It looks stable enough ( with documentation available)

If a row exist in CrawlParticipants, an invitation is sent. Status is then a boolean stating whether or not you have accepted.

We need to figure what data we should save from facebook and store in facebook-table, and what data we should save in user.

Part of this discussion will be if we require facebook-account, or if we will support facebook autentication.

Edit:
https://developers.facebook.com/tools/explorer?method=GET&path=me%3Ffields%3Dlocation%2Cname%2Cemail%2Cpicture%2Ccover&version=v2.8

**Review by magnumog — APPROVED** [review]
Burde id i tabellene bli definert ved hjelp av UUID (string) istedenfor INT (løpenummer)?

**Comment by johnolos** [comment]
It's generally advised against using GUIDs as primary key due to the space overhead. 128 bit vs 32 bit.
http://www.sqlskills.com/blogs/kimberly/guids-as-primary-keys-andor-the-clustering-key/

**Comment by hessenh** [comment]
Good work! 
I don't think it's necessary to have a table for Facebook-users. What do you think @hawkaa ?

**Comment by magnumog** [comment]
Nice, did not think about the overhead and the speed benefits og having the numbers sequential

File: src/main/scala/io/pubcrawler/stout/db/Schemas.scala [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  def routeId = column[Int]("ROUTE_STOP_ID")
+  def stopId = column[Int]("STOP_ID")
+  def order = column[Int]("ORDER")
+  def * = (routeId, stopId, order)
+
+  def stop = foreignKey("STOP_FK", stopId, stops)(_.id)
+  def route = foreignKey("ROUTE_FK", routeId, routes)(_.id)
+}
+
+class Crawl(tag: Tag, stops: TableQuery[Stop]) extends Table[(Int, String, Int, Time, Float, Float, Int, String, String)](tag, "CRAWL") {
```
Comment (johnolos) [inline src/main/scala/io/pubcrawler/stout/db/Schemas.scala]: stops are not needed here.

File: src/main/scala/io/pubcrawler/stout/db/Schemas.scala [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  def name = column[String]("NAME")
+  def birthdate = column[Date]("BIRTHDATE")
+  def location = column[String]("LOCATION")
+  def gender = column[String]("GENDER")
+  def facebookId = column[Int]("FACEBOOK_ID")
+  def * = (id, name, birthdate, location, gender, facebookId)
+}
+
+class Wishes(tag: Tag, stops: TableQuery[Stop], users: TableQuery[User]) extends Table[(Int, Int, Int)](tag, "WISHES") {
+  def id = column[Int]("WISHES", O.PrimaryKey)
```
Comment (johnolos) [inline src/main/scala/io/pubcrawler/stout/db/Schemas.scala]: `WISH_ID`

File: src/main/scala/io/pubcrawler/stout/db/Schemas.scala [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  def id = column[Int]("FACEBOOK_USER_ID", O.PrimaryKey)
+  def name = column[String]("NAME")
+  def birthdate = column[Date]("BIRTHDATE")
+  def location = column[String]("LOCATION")
+  def gender = column[String]("GENDER")
+  def facebookId = column[Int]("FACEBOOK_ID")
+  def * = (id, name, birthdate, location, gender, facebookId)
+}
+
+class Wishes(tag: Tag, stops: TableQuery[Stop], users: TableQuery[User]) extends Table[(Int, Int, Int)](tag, "WISHES") {
```
Comment (johnolos) [inline src/main/scala/io/pubcrawler/stout/db/Schemas.scala]: `Wish`: Better to keep it singular form.

**Review by johnolos — COMMENTED** [review]
Few notes to myself.

File: src/main/scala/io/pubcrawler/stout/db/Schemas.scala [hunk 4]
```diff
@@ -0,0 +1,82 @@
+package io.pubcrawler.stout.db
+
+import slick.jdbc.PostgresProfile.api._
+import java.sql.Date
+import java.sql.Time
+
+
+class Stop(tag: Tag) extends Table[(Int, String, String, Float, Float)](tag, "STOP") {
```
Comment (hawkaa) [inline src/main/scala/io/pubcrawler/stout/db/Schemas.scala]: I think the `Table` subclasses should have Plural naming, which would be `Stops` in this case.

Any chance we can throw in some case classes to represent instances? http://slick.lightbend.com/doc/3.1.1/schemas.html#mapped-tables

File: src/main/scala/io/pubcrawler/stout/db/Schemas.scala [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  def id = column[Int]("STOP_ID", O.PrimaryKey)
```
Comment (hawkaa) [inline src/main/scala/io/pubcrawler/stout/db/Schemas.scala]: Why uppercase?

File: build.sbt [hunk 6]
```diff
@@ -24,8 +24,11 @@ libraryDependencies ++= Seq(
   "org.scalactic" %% "scalactic" % "3.0.1",
   "org.scalatest" %% "scalatest" % "3.0.1" % "test",
   "ch.qos.logback" % "logback-classic" % "1.1.7" % "runtime",
-  "org.eclipse.jetty" % "jetty-webapp" % "9.2.15.v20160210" % "container",
   "com.typesafe.scala-logging" %% "scala-logging" % "3.5.0",
+  "com.typesafe.slick" %% "slick" % "3.2.0-M2",
```
Comment (hawkaa) [inline build.sbt:28]: I'm glad you pull in the Scala 2.12 version :)

File: src/main/scala/io/pubcrawler/stout/db/Schemas.scala [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  def username = column[String]("USERNAME")
+  def birthdate = column[Date]("BIRTHDATE")
+  def gender = column[String]("GENDER")
+  def facebookUserId = column[Int]("FACEBOOK_ID")
+  def * = (id, username, birthdate, gender, facebookUserId)
+
+  def facebookUser = foreignKey("FACEBOOK_USER_FK", facebookUserId, facebookUsers)(_.id)
+}
+
+class FacebookUser(tag: Tag) extends Table[(Int, String, Date, String, String, Int)](tag, "FACEBOOK_USER") {
```
Comment (hawkaa) [inline src/main/scala/io/pubcrawler/stout/db/Schemas.scala]: I never meant to have a separate "Facebook user" store in the database, I just wanted to add information in the database diagram that we could fetch info from the facebook user information. Please remove.

File: src/main/scala/io/pubcrawler/stout/db/Schemas.scala (hunk 3 above)
Comment (hawkaa) [reply src/main/scala/io/pubcrawler/stout/db/Schemas.scala]: Seems like convention is plural.

File: build.sbt [hunk 8]
```diff
@@ -24,8 +24,11 @@ libraryDependencies ++= Seq(
...
+  "com.typesafe.slick" %% "slick-hikaricp" % "3.2.0-M2",
+  "org.eclipse.jetty" % "jetty-webapp" % "9.2.15.v20160210" % "container",
+  "org.slf4j" % "slf4j-nop" % "1.6.4",
```
Comment (hawkaa) [inline build.sbt]: Consider adding in the `simple` logger rather than the `nop`.

File: src/main/scala/io/pubcrawler/stout/db/Schemas.scala [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  def radius = column[Int]("RADIUS")
+  def city = column[String]("CITY")
+  def description = column[String]("DESCRIPTION")
+  def * = (id, title, ownerId, dateTime, lat, lng, radius, city, description)
+}
+
+class CrawlParticipants(tag: Tag) extends Table[(Int, Int, Boolean)](tag, "CRAWL_PARTICIPANTS") {
+  def crawlId = column[Int]("CRAWL_ID")
+  def userId = column[Int]("USER_ID")
+  def status = column[Boolean]("STATUS")
```
Comment (hawkaa) [inline src/main/scala/io/pubcrawler/stout/db/Schemas.scala]: Why not enum? Invited, accepted, declined?

**Comment by hawkaa** [comment]
Good job setting this up. I have some comments:

- Remove the FacebookUser, that was just to represent that we had an external data source
- Convention seems to have the `Table` subclasses plural, not singular
- Lowercase instead of uppercase for table columns?
- Consider case classes to represent table elements (http://slick.lightbend.com/doc/3.1.1/schemas.html#mapped-tables)

File: src/main/scala/io/pubcrawler/stout/db/Schemas.scala (hunk 7 above)
Comment (johnolos) [reply src/main/scala/io/pubcrawler/stout/db/Schemas.scala]: 👍

File: src/main/scala/io/pubcrawler/stout/db/Schemas.scala (hunk 9 above)
Comment (johnolos) [reply src/main/scala/io/pubcrawler/stout/db/Schemas.scala]: I was taking notes from the docs, but I'll look into enums, and I wholeheartedly agree it reads better with enums.

http://slick.lightbend.com/doc/3.1.1/schemas.html#table-rows

Comment (johnolos) [reply src/main/scala/io/pubcrawler/stout/db/Schemas.scala]: I discussed this briefly with Hans-Olav and we both agree that if there exists a row in CrawlParticipants, you're considered to be Invited. I'll change this to Enum with Accepted, Maybe, Declined.

**Comment by johnolos** [comment]
@hawkaa Does it look roughly OK now? :-)

**Comment by hawkaa** [comment]
Looks good 👍

**Comment by johnolos** [comment]
There will be rounds with db PRs. I think it's just fair enough to merge this PR and take one change at a time. :-)

**Pull request merged**